packages/browser-agent-core/common/wrap/wrap-promise.js

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright 2020 New Relic Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+import {createWrapperWithEmitter as wrapFn, wrapInPlace, argsToArray} from './wrap-function'
+import {ee as baseEE, getOrSetContext} from '../event-emitter/contextual-ee'
+import {mapOwn} from '../util/map-own'
+import {originals} from '../config/config'
+
+const wrapped = {}
+
+export function wrapPromise(sharedEE){
+  const promiseEE = scopedEE(sharedEE)
+  if (wrapped[promiseEE.debugId]) return promiseEE
+  wrapped[promiseEE.debugId] = true
+  var getContext = getOrSetContext
+  var promiseWrapper = wrapFn(promiseEE)
+  var OriginalPromise = originals.PR
+  
+  if (OriginalPromise) {
+    wrap()
+  }
+  
+  function wrap() {
+    window.Promise = WrappedPromise
+  
+    ;['all', 'race'].forEach(function (method) {
+      var original = OriginalPromise[method]
+      OriginalPromise[method] = function (subPromises) {
+        var finalized = false
+        mapOwn(subPromises, function (i, sub) {
+          // eslint-disable-next-line
+          Promise.resolve(sub).then(setNrId(method === 'all'), setNrId(false))
+        })
+  
+        var originalReturnValue = original.apply(OriginalPromise, arguments)
+        var promise = OriginalPromise.resolve(originalReturnValue)
+  
+        return promise
+  
+        function setNrId (overwrite) {
+          return function () {
+            promiseEE.emit('propagate', [null, !finalized], originalReturnValue, false, false)
+            finalized = finalized || !overwrite
+          }
+        }
+      }
+    })
+  
+    ;['resolve', 'reject'].forEach(function (method) {
+      var original = OriginalPromise[method]
+      OriginalPromise[method] = function (val) {
+        var returnVal = original.apply(OriginalPromise, arguments)
+        if (val !== returnVal) {
+          promiseEE.emit('propagate', [val, true], returnVal, false, false)
+        }
+  
+        return returnVal
+      }
+    })
+  
+    OriginalPromise.prototype['catch'] = function wrappedCatch (fn) {
+      return this.then(null, fn)
+    }
+  
+    Object.assign(OriginalPromise.prototype, {constructor: {value: WrappedPromise}})
```

Comment:
Why was this changed?

Reply:
I believe this was added by @cwli24 for the circular interaction ticket?